### PR TITLE
Merge `Between` and `VariableDiff` with `RelativePoseError` and `PosePriorError`

### PR DESCRIPTION
**Pull request by fantaosha**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by mhmukadam — COMMENTED** [review]
Looks good, only one comment.

File: theseus/geometry/se3.py [hunk 1]
```diff
@@ -22,12 +22,13 @@ def __init__(
...
         name: Optional[str] = None,
         dtype: Optional[torch.dtype] = None,
+        requires_check: bool = True,
     ):
         if x_y_z_quaternion is not None and data is not None:
             raise ValueError("Please provide only one of x_y_z_quaternion or data.")
         if x_y_z_quaternion is not None:
             dtype = x_y_z_quaternion.dtype
-        if data is not None:
+        if data is not None and requires_check:
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/geometry/se3.py:31]: Yes, if `data` is used for initialization, the default `dtype` will be `data.dtype`.

[the remark this answers is not in the document]
Comment (mhmukadam) [reply theseus/geometry/se3.py:31]: Ah yes super goes all the way back to `Variable` where it is assigned. Sounds good.

File: examples/pose_graph.py [hunk 2]
```diff
@@ -12,37 +12,26 @@
...
+    cost_func = th.eb.Between(
+        verts[edge.i], verts[edge.j], edge.weight, edge.relative_pose
     )
     objective.add(cost_func)
 
-objective.add(
-    theg.PosePriorError(
-        pose=verts[0], pose_prior=verts[0].copy(new_name=verts[0].name + "PRIOR")
-    )
+pose_prior = th.eb.VariableDifference(
```
Comment (luisenp) [inline examples/pose_graph.py:29]: I'm confused about this cost function. Is this a constraint to make the initial pose exactly some value? Why is the weight so small?

File: theseus/utils/examples/pose_graph/dataset.py [hunk 3]
```diff
@@ -47,7 +49,7 @@ def read_3D_g2o_file(path: str) -> Tuple[int, List[th.SE3], List[PoseGraphEdge]]
 
                 x_y_z_quat = torch.from_numpy(
                     np.array([tokens[3:10]], dtype=np.float64)
-                )
+                ).to(dtype)
```
Comment (luisenp) [inline theseus/utils/examples/pose_graph/dataset.py:52]: Why not change the line above to do `, dtype=dtype)`?

File: theseus/utils/examples/pose_graph/dataset.py [hunk 4]
```diff
@@ -56,6 +58,7 @@ def read_3D_g2o_file(path: str) -> Tuple[int, List[th.SE3], List[PoseGraphEdge]]
                 sel = [0, 6, 11, 15, 18, 20]
                 weight = th.Variable(
                     torch.from_numpy(np.array(tokens[10:], dtype=np.float64)[sel])
+                    .to(dtype)
```
Comment (luisenp) [inline theseus/utils/examples/pose_graph/dataset.py:61]: Same comment about `, dtype=dtype)`

File: theseus/utils/examples/pose_graph/dataset.py [hunk 5]
```diff
@@ -74,7 +77,9 @@ def read_3D_g2o_file(path: str) -> Tuple[int, List[th.SE3], List[PoseGraphEdge]]
             elif tokens[0] == "VERTEX_SE3:QUAT":
                 i = int(tokens[1])
 
-                x_y_z_quat = torch.from_numpy(np.array([tokens[2:]], dtype=np.float64))
+                x_y_z_quat = torch.from_numpy(
+                    np.array([tokens[2:]], dtype=np.float64)
+                ).to(dtype)
```
Comment (luisenp) [inline theseus/utils/examples/pose_graph/dataset.py:82]: Same comment about `, dtype=dtype)`

File: theseus/utils/examples/pose_graph/dataset.py [hunk 6]
```diff
@@ -113,14 +120,17 @@ def read_2D_g2o_file(path: str) -> Tuple[int, List[th.SE2], List[PoseGraphEdge]]
 
                 n = len(edges)
 
-                x_y_theta = torch.from_numpy(np.array([tokens[3:6]], dtype=np.float64))
+                x_y_theta = torch.from_numpy(
+                    np.array([tokens[3:6]], dtype=np.float64)
```
Comment (luisenp) [inline theseus/utils/examples/pose_graph/dataset.py:124]: Same comment about `, dtype=dtype)`

File: theseus/utils/examples/pose_graph/dataset.py [hunk 7]
```diff
@@ -113,14 +120,17 @@ def read_2D_g2o_file(path: str) -> Tuple[int, List[th.SE2], List[PoseGraphEdge]]
...
+                ).to(dtype)
                 relative_pose = th.SE2(
                     x_y_theta=x_y_theta, name="EDGE_SE2__{}".format(n)
                 )
 
                 sel = [0, 3, 5]
                 weight = th.Variable(
                     torch.from_numpy(np.array(tokens[6:], dtype=np.float64)[sel])
+                    .to(dtype)
```
Comment (luisenp) [inline theseus/utils/examples/pose_graph/dataset.py:133]: Same comment about `, dtype=dtype)`

File: theseus/utils/examples/pose_graph/dataset.py [hunk 8]
```diff
@@ -139,7 +149,9 @@ def read_2D_g2o_file(path: str) -> Tuple[int, List[th.SE2], List[PoseGraphEdge]]
             elif tokens[0] == "VERTEX_SE2:QUAT":
                 i = int(tokens[1])
 
-                x_y_theta = torch.from_numpy(np.array([tokens[2:]], dtype=np.float64))
+                x_y_theta = torch.from_numpy(
```
Comment (luisenp) [inline theseus/utils/examples/pose_graph/dataset.py:152]: Same comment about `, dtype=dtype)`

**Review by luisenp — COMMENTED** [review]
Looks good! Just left a few comments and a question.

File: theseus/utils/examples/pose_graph/dataset.py (hunk 3 above)
Comment (fantaosha) [reply theseus/utils/examples/pose_graph/dataset.py:52]: It looks like [`torch.from_numpy`](https://pytorch.org/docs/stable/generated/torch.from_numpy.html) has no `dtype` options.

File: examples/pose_graph.py (hunk 2 above)
Comment (fantaosha) [reply examples/pose_graph.py:29]: Correct. The only reason we introduce `pose_prior` is to make the linear system not ill-conditioned. A too large weight will restrict the search space and yield slow convergence.

Comment (fantaosha) [reply examples/pose_graph.py:29]: `GTSAM` also has similar factors [[link](https://github.com/borglab/gtsam/blob/0850e89b0e6148d16f00b7741d83b6325ada977d/examples/Pose2SLAMExample.cpp#L73)].

Comment (luisenp) [reply examples/pose_graph.py:29]: If I understand that code, they are anchoring that variable to a zero/origin value, which is not the case here? Here the target is set to whatever value `verts[0]` has at this point.

File: theseus/utils/examples/pose_graph/dataset.py (hunk 3 above)
Comment (luisenp) [reply theseus/utils/examples/pose_graph/dataset.py:52]: But there is a `dtype` inside the `np.array` which we can use. I think torch will respect the type of the underlying numpy array.

Comment (fantaosha) [reply theseus/utils/examples/pose_graph/dataset.py:52]: I know. But the `dtype` in `np.array()` does not accept types like `torch.float32`.

File: examples/pose_graph.py (hunk 2 above)
Comment (fantaosha) [reply examples/pose_graph.py:29]: I don't think the `target` value matters. In fact, the `target` value can be anything. In PGO problems, we mainly consider the relative pose errors, which, however, has no constraints on the absolute pose. Thus, we introduce pose prior add absolute pose constraint.

The reason why I use `verts[0]` is that it is closed to the initial guess.

File: theseus/utils/examples/pose_graph/dataset.py (hunk 3 above)
Comment (luisenp) [reply theseus/utils/examples/pose_graph/dataset.py:52]: Why can't you do something like this

image

File: examples/pose_graph.py (hunk 2 above)
Comment (luisenp) [reply examples/pose_graph.py:29]: I see. Thanks for the explanation!

File: theseus/utils/examples/pose_graph/dataset.py (hunk 3 above)
Comment (fantaosha) [reply theseus/utils/examples/pose_graph/dataset.py:52]: Yes, I can do this. But `dtype` is one the input of [`read_3D_g2o_file`](https://github.com/facebookresearch/theseus/blob/bed363f5c1643ffa146c57d0ddde80dd5e28fc5b/theseus/utils/examples/pose_graph/dataset.py#L32). Is it OK to mix the use of `torch.dtype` and `np.dtype`?

**Review by mhmukadam — APPROVED** [review]

**Pull request merged**